File: backend/app/schemas/job_training.py
"""Pydantic schemas for the STUDENT side of Job Training
(database/migrations/052_job_training.sql).

PHASE J3 scope: a SELECTED job candidate whose industry has authored a
job training program can list their enrollment(s) and read the PUBLISHED
program (job info + program metadata + published modules / items /
assignments + skills). No completion, certificate, progress, submission,
or notification surface -- those are later phases.

There are NO request bodies here: both student endpoints are GETs whose
identity comes entirely from the token. The one non-GET surface (the
industry self-heal endpoint) carries no body either. `ProvisionJobTrainingResponse`
is the shape returned by that industry-side heal endpoint.
"""

from pydantic import BaseModel


class JobTrainingEnrollmentSummary(BaseModel):
    """One row of the student's job training list. `program_*` are null
    while the program is still DRAFT (the student cannot read an
    unpublished program); `job_title` is best-effort (a CLOSED / ARCHIVED
    job is not student-readable, which never affects access)."""

    enrollment_id: str
    application_id: str
    job_id: str
    job_title: str | None = None
    program_id: str | None = None
    program_title: str | None = None
    program_status: str | None = None
    enrollment_status: str  # ACTIVE | COMPLETED  (REVOKED is never listed)
    created_at: str | None = None
    completed_at: str | None = None


class JobTrainingListResponse(BaseModel):
    enrollments: list[JobTrainingEnrollmentSummary]


class StudentJobProgramItem(BaseModel):
    id: str
    title: str
    item_type: str
    content_url: str | None = None
    content_text: str | None = None
    order_index: int


class StudentJobProgramAssignment(BaseModel):
    id: str
    title: str
    description: str | None = None
    instructions: str | None = None
    assignment_type: str
    is_required: bool
    order_index: int
    due_offset_days: int | None = None
    submission_kind: str
    repo_required: bool
    live_url_expected: bool
    max_score: float | None = None
    linked_skill_id: str | None = None


class StudentJobProgramModule(BaseModel):
    id: str
    title: str
    description: str | None = None
    order_index: int
    items: list[StudentJobProgramItem] = []
    assignments: list[StudentJobProgramAssignment] = []


class StudentJobProgramSkill(BaseModel):
    skill_id: str
    skill_name: str
    requirement: str  # REQUIRED | OPTIONAL


class StudentJobProgramMeta(BaseModel):
    id: str
    job_id: str
    title: str
    summary: str | None = None
    estimated_weeks: int | None = None
    status: str  # always PUBLISHED in a successful detail response
    published_at: str | None = None


class StudentJobTrainingDetail(BaseModel):
    enrollment: JobTrainingEnrollmentSummary
    program: StudentJobProgramMeta
    modules: list[StudentJobProgramModule] = []
    skills: list[StudentJobProgramSkill] = []


# ============================================================
# industry self-heal endpoint response
# ============================================================


class JobTrainingEnrollmentRef(BaseModel):
    """The raw enrollment row, as returned to the OWNING INDUSTRY from the
    self-heal endpoint (they own the posting and legitimately see the
    applicant). Never returned to a student."""

    id: str
    application_id: str
    job_id: str
    student_id: str
    industry_id: str
    enrollment_status: str
    created_at: str | None = None
    completed_at: str | None = None
    revoked_at: str | None = None


class ProvisionJobTrainingResponse(BaseModel):
    outcome: str  # CREATED | ALREADY_EXISTS | SKIPPED_NO_PROGRAM |
    #               SKIPPED_NOT_SELECTED | SKIPPED_NOT_JOB | REVOKED_BLOCKED
    detail: str
    enrollment: JobTrainingEnrollmentRef | None = None

File: frontend/types/job-training.ts
/**
 * Mirrors backend/app/schemas/job_training.py -- field-for-field, same
 * nullability.
 *
 * Job Training is the post-selection experience for a student SELECTED for
 * a JOB whose industry has authored a training program
 * (database/migrations/052_job_training.sql). It is COMPLETELY separate
 * from:
 *   - the universal Learning & Courses catalog (/student/learning), and
 *   - the Internship Workspace (/student/my-internships), which is what a
 *     SELECTED *internship* candidate uses.
 *
 * The authoritative signal that a student HAS Job Training is a
 * non-revoked enrollment returned by GET /api/v1/student/job-training --
 * never `application.status === "SELECTED"` on its own (a selected job
 * with no authored program produces no enrollment).
 */

export type JobTrainingEnrollmentStatus = "ACTIVE" | "COMPLETED" | "REVOKED";

/** Friendly label -- the backend value is never changed, only presented.
 * REVOKED is never surfaced (the list omits it, the detail 404s). */
export const JOB_TRAINING_ENROLLMENT_STATUS_LABEL: Record<
  JobTrainingEnrollmentStatus,
  string
> = {
  ACTIVE: "Active",
  COMPLETED: "Completed",
  REVOKED: "Revoked",
};

export interface JobTrainingEnrollmentSummary {
  enrollment_id: string;
  application_id: string;
  job_id: string;
  /** Best-effort: null when the job posting is no longer student-readable
   * (CLOSED / ARCHIVED). This never affects access. */
  job_title: string | null;
  /** null while the program is still a DRAFT (a student cannot read an
   * unpublished program). */
  program_id: string | null;
  program_title: string | null;
  program_status: string | null;
  enrollment_status: string;
  created_at: string | null;
  completed_at: string | null;
}

export interface JobTrainingListResponse {
  enrollments: JobTrainingEnrollmentSummary[];
}

export type JobTrainingItemType = "VIDEO" | "PDF" | "LINK" | "TEXT";

export interface StudentJobProgramItem {
  id: string;
  title: string;
  item_type: string;
  content_url: string | null;
  content_text: string | null;
  order_index: number;
}

export interface StudentJobProgramAssignment {
  id: string;
  title: string;
  description: string | null;
  instructions: string | null;
  assignment_type: string;
  is_required: boolean;
  order_index: number;
  due_offset_days: number | null;
  submission_kind: string;
  repo_required: boolean;
  live_url_expected: boolean;
  max_score: number | null;
  linked_skill_id: string | null;
}

export interface StudentJobProgramModule {
  id: string;
  title: string;
  description: string | null;
  order_index: number;
  items: StudentJobProgramItem[];
  assignments: StudentJobProgramAssignment[];
}

export interface StudentJobProgramSkill {
  skill_id: string;
  skill_name: string;
  requirement: "REQUIRED" | "OPTIONAL";
}

export interface StudentJobProgramMeta {
  id: string;
  job_id: string;
  title: string;
  summary: string | null;
  estimated_weeks: number | null;
  /** Always "PUBLISHED" in a successful detail response. */
  status: string;
  published_at: string | null;
}

export interface StudentJobTrainingDetail {
  enrollment: JobTrainingEnrollmentSummary;
  program: StudentJobProgramMeta;
  modules: StudentJobProgramModule[];
  skills: StudentJobProgramSkill[];
}
